File: Baekjoon/Class3/31.1389 케빈 베이컨의 6단계 법칙.py
#너비 우선 탐색
from collections import deque
import sys
input = sys.stdin.readline

n, m = map(int, input().split())

graph = [[] for _ in range(n + 1)] #무방향성 그래프
for _ in range(m):
    start, end = map(int, input().split())
    graph[start].append(end)
    graph[end].append(start)

min_kevin = float("inf")
min_kevin_idx = 0
for i in range(1, n + 1):
    kevin = 0
    bfs = deque() #너비 우선 탐색
    depths = [-1] * (n + 1)
    bfs.append(i)
    depths[i] = 0
    while bfs:
        start = bfs.popleft()
        for end in graph[start]:
            if depths[end] == -1:
                bfs.append(end)
                depths[end] = depths[start] + 1
                kevin += depths[end] #케빈 베이컨 수 누적
    if min_kevin > kevin:
        min_kevin = kevin
        min_kevin_idx = i

print(min_kevin_idx)

''' 플로이드-워셜
import sys
input = sys.stdin.readline

MAX_USERS = 100

n, m = map(int, input().split())

graph = [[MAX_USERS] * (n + 1) for _ in range(n + 1)] #격자 그래프
for _ in range(m):
    start, end = map(int, input().split())
    graph[start][end] = 1
    graph[end][start] = 1

for k in range(1, n + 1):
    for i in range(1, n + 1):
        for j in range(1, n + 1):
            graph[i][j] = min(graph[i][j], graph[i][k] + graph[k][j]) #지름길 찾기

for i in range(1, n + 1): #자기 자신에 대한 경로 0
    graph[i][i] = 0

min_kevin = float("inf")
min_kevin_idx = 0
for i in range(1, n + 1):
    kevin = sum(graph[i][1:]) #케빈 베이컨 수 누적
    if min_kevin > kevin:
        min_kevin = kevin
        min_kevin_idx = i

print(min_kevin_idx)
'''

File: Baekjoon/Class3/31.1389 케빈 베이컨의 6단계 법칙.js
// 너비 우선 탐색
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split("\n");

const [n, m] = input[0].split(" ").map(Number);

const graph = Array.from({ length : n + 1 }, () => []); // 무방향성 그래프
for(let i = 1; i <= m; i++) {
    const [start, end] = input[i].split(" ").map(Number);
    graph[start].push(end);
    graph[end].push(start);
}

let minKevin = Number.MAX_VALUE;
let minKevinIdx = 0;
for(let i = 1; i <= n; i++) {
    let kevin = 0;
    let bfsHead = 0;
    const bfs = []; // 너비 우선 탐색
    const depths = new Array(n + 1).fill(-1);
    bfs.push(i);
    depths[i] = 0;
    while(bfsHead < bfs.length) {
        const start = bfs[bfsHead++];
        for(const end of graph[start]) {
            if(depths[end] === -1) {
                bfs.push(end);
                depths[end] = depths[start] + 1;
                kevin += depths[end]; // 케빈 베이컨 수 누적
            }
        }
    }
    if(minKevin > kevin) {
        minKevin = kevin;
        minKevinIdx = i;
    }
}

console.log(minKevinIdx);


/* 플로이드-워셜
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split("\n");

const MAX_USERS = 100;

const [n, m] = input[0].split(" ").map(Number);

const graph = Array.from({ length : n + 1 }, () => new Array(n + 1).fill(MAX_USERS));
for(let i = 1; i <= m; i++) {
    const [start, end] = input[i].split(" ").map(Number);
    graph[start][end] = 1;
    graph[end][start] = 1;
}

for(let k = 1; k <= n; k++) // 모든 쌍 최단 경로
    for(let i = 1; i <= n; i++)
        for(let j = 1; j <= n; j++)
            graph[i][j] = Math.min(graph[i][j], graph[i][k] + graph[k][j]); // 지름길 찾기

for(let i = 1; i <= n; i++) // 자기 자신에 대한 경로 0
    graph[i][i] = 0;

let minKevin = Number.MAX_VALUE;
let minKevinIdx = 0;
for(let i = 1; i <= n; i++) {
    const kevin = graph[i].slice(1).reduce((acc, cur) => acc + cur); // 케빈 베이컨 수 누적
    if(minKevin > kevin) {
        minKevin = kevin;
        minKevinIdx = i;
    }
}
console.log(minKevinIdx);
*/